### PR TITLE
Add Route 53 Resolver endpoint resource

### PR DESCRIPTION
**Pull request by ewbankkit**
Fixes https://github.com/terraform-providers/terraform-provider-aws/issues/6563.
Includes:
* https://github.com/terraform-providers/terraform-provider-aws/pull/6549
* https://github.com/terraform-providers/terraform-provider-aws/pull/6554

Acceptance tests (so far):

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsRoute53ResolverEndpoint_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsRoute53ResolverEndpoint_ -timeout 120m
=== RUN   TestAccAwsRoute53ResolverEndpoint_importBasic
=== PAUSE TestAccAwsRoute53ResolverEndpoint_importBasic
=== RUN   TestAccAwsRoute53ResolverEndpoint_basicInbound
=== PAUSE TestAccAwsRoute53ResolverEndpoint_basicInbound
=== RUN   TestAccAwsRoute53ResolverEndpoint_updateOutbound
=== PAUSE TestAccAwsRoute53ResolverEndpoint_updateOutbound
=== CONT  TestAccAwsRoute53ResolverEndpoint_importBasic
=== CONT  TestAccAwsRoute53ResolverEndpoint_updateOutbound
=== CONT  TestAccAwsRoute53ResolverEndpoint_basicInbound
--- PASS: TestAccAwsRoute53ResolverEndpoint_importBasic (108.99s)
--- PASS: TestAccAwsRoute53ResolverEndpoint_basicInbound (129.35s)
--- PASS: TestAccAwsRoute53ResolverEndpoint_updateOutbound (503.80s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	503.819s
```

### CODE REVIEW

**Comment by gazoakley** [comment]
You might be able to use `TypeList` with `DiffSuppressFunc` and match things  up by `ip_id`. That sounds like a pain, but it might work.

**Comment by gazoakley** [comment]
Your `DiffSuppressFunc` might:

- Check the `ip_address` entries where `ip` is specified all exist and are associated with the correct `subnet_id`
- Check there are the correct number of entries per `subnet_id` for the remaining `ip_address` entries
- Suppress the diff if all the above pass

**Comment by ewbankkit** [comment]
Not having much luck with the `ip_address` diff issue and won't have time to work on this for the next week or so.
I've reverted to a set for now with still the same concerns:

```go
	// TODO
	// TODO * "ip" may be computed - Blank on input, set by API on output; How to handle?
	// TODO * Multiple "ip_address" values per "subnet_id" are allowed in the API
	// TODO   but a List (instead of a Set) won't work because of non-deterministic order
	// TODO   on listing ip_address, plus how to determine diffs for update?
	// TODO
```

Acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsRoute53ResolverEndpoint_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsRoute53ResolverEndpoint_ -timeout 120m
=== RUN   TestAccAwsRoute53ResolverEndpoint_basicInbound
=== PAUSE TestAccAwsRoute53ResolverEndpoint_basicInbound
=== RUN   TestAccAwsRoute53ResolverEndpoint_updateOutbound
=== PAUSE TestAccAwsRoute53ResolverEndpoint_updateOutbound
=== CONT  TestAccAwsRoute53ResolverEndpoint_basicInbound
=== CONT  TestAccAwsRoute53ResolverEndpoint_updateOutbound
--- PASS: TestAccAwsRoute53ResolverEndpoint_basicInbound (135.30s)
--- PASS: TestAccAwsRoute53ResolverEndpoint_updateOutbound (516.18s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	516.201s
```

**Comment by ewbankkit** [comment]
Rebased to fix conflicts.
Re-ran acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsRoute53ResolverEndpoint_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsRoute53ResolverEndpoint_ -timeout 120m
=== RUN   TestAccAwsRoute53ResolverEndpoint_basicInbound
=== PAUSE TestAccAwsRoute53ResolverEndpoint_basicInbound
=== RUN   TestAccAwsRoute53ResolverEndpoint_updateOutbound
=== PAUSE TestAccAwsRoute53ResolverEndpoint_updateOutbound
=== CONT  TestAccAwsRoute53ResolverEndpoint_basicInbound
=== CONT  TestAccAwsRoute53ResolverEndpoint_updateOutbound
--- PASS: TestAccAwsRoute53ResolverEndpoint_basicInbound (132.30s)
--- PASS: TestAccAwsRoute53ResolverEndpoint_updateOutbound (569.06s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	569.081s
```

**Comment by ewbankkit** [comment]
@bflad Looking for advice on the best way of dealing with the set/list conundrum for the `ip_address` attribute. It best fits the requirements for a set - Order is unimportant, but there are problems with determining the correct hash function.

**Comment by ewbankkit** [comment]
TODO: Update tag handling code to reflect comments [here](https://github.com/terraform-providers/terraform-provider-aws/pull/6911#discussion_r242696623).

**Comment by rbadillo** [comment]
Hi team, any updates when this will be available ?

**Comment by rbadillo** [comment]
Hi team, any updates when this will be available ?

@ewbankkit

**Comment by ewbankkit** [comment]
@bflad Any thoughts on my stumbling block - https://github.com/terraform-providers/terraform-provider-aws/pull/6574#issuecomment-447685757?

**Comment by aar6ncai** [comment]
hi guys, any update for this feature become available ? thanks

**Comment by ewbankkit** [comment]
@aar6ncai I'm going to need this quite soon too 😄.
I'll remove _WIP_ and submit for formal review with the caveat that there is the known concern around `ip` in the `ip_address` block.

**Comment by kmcquade** [comment]
@ewbankkit @gazoakley - any word? We need this pretty soon as well.

**Comment by ewbankkit** [comment]
* Rebased to remove conflict
* Correct acceptance tests to use `tags = {}` syntax for TF 0.12

Acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsRoute53ResolverEndpoint_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsRoute53ResolverEndpoint_ -timeout 120m
=== RUN   TestAccAwsRoute53ResolverEndpoint_basicInbound
=== PAUSE TestAccAwsRoute53ResolverEndpoint_basicInbound
=== RUN   TestAccAwsRoute53ResolverEndpoint_updateOutbound
=== PAUSE TestAccAwsRoute53ResolverEndpoint_updateOutbound
=== CONT  TestAccAwsRoute53ResolverEndpoint_basicInbound
=== CONT  TestAccAwsRoute53ResolverEndpoint_updateOutbound
--- PASS: TestAccAwsRoute53ResolverEndpoint_basicInbound (108.17s)
--- PASS: TestAccAwsRoute53ResolverEndpoint_updateOutbound (614.07s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	629.443s
```

**Comment by ewbankkit** [comment]
Added sweeper (otherwise dangling resources will give you a nasty surprise on your AWS bill 😄).

```console
$ go test ./aws -v -sweep=us-west-2 -sweep-run=aws_route53_resolver_endpoint -timeout 10h
2019/01/25 18:34:13 [DEBUG] Running Sweepers for region (us-west-2):
2019/01/25 18:34:13 [INFO] Building AWS region structure
2019/01/25 18:34:13 [INFO] Building AWS auth structure
2019/01/25 18:34:13 [INFO] Setting AWS metadata API timeout to 100ms
2019/01/25 18:34:13 [INFO] Ignoring AWS metadata API endpoint at default location as it doesn't return any instance-id
2019/01/25 18:34:13 [INFO] AWS Auth provider used: "EnvProvider"
2019/01/25 18:34:13 [INFO] Initializing DeviceFarm SDK connection
2019/01/25 18:34:13 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2019/01/25 18:34:14 Sweeper Tests ran:
	- aws_route53_resolver_endpoint
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1.862s
```

**Comment by kmcquade** [comment]
@bflad any updates? thanks.
CC: @ewbankkit

**Comment by ewbankkit** [comment]
Rebased to resolve conflicts.

**Comment by ewbankkit** [comment]
Removed sweeper prefix check (https://github.com/terraform-providers/terraform-provider-aws/pull/7484).

File: aws/resource_aws_route53_resolver_endpoint.go [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+}
+
+func route53ResolverHashIPAddress(v interface{}) int {
+	var buf bytes.Buffer
+	m := v.(map[string]interface{})
+	buf.WriteString(fmt.Sprintf("%s-", m["subnet_id"].(string)))
+	// TODO
+	// TODO * "ip" may be computed - Blank on input, set by API on output; How to handle?
+	// TODO * Multiple "ip_address" values per "subnet_id" are allowed in the API
+	// TODO   but a List (instead of a Set) won't work because of non-deterministic order
```
Comment (paultyng) [inline aws/resource_aws_route53_resolver_endpoint.go]: Could you just always sort prior to hashing?

File: aws/resource_aws_route53_resolver_endpoint.go [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/route53resolver"
+	"github.com/hashicorp/terraform/helper/hashcode"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+const (
+	Route53ResolverEndpointStatusDeleted = "DELETED"
```
Comment (paultyng) [inline aws/resource_aws_route53_resolver_endpoint.go]: this probably doesn't need to be exported

Comment (ewbankkit) [reply aws/resource_aws_route53_resolver_endpoint.go]: correct, it doesn't

**Comment by ewbankkit** [comment]
Re-ran acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsRoute53ResolverEndpoint_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsRoute53ResolverEndpoint_ -timeout 120m
=== RUN   TestAccAwsRoute53ResolverEndpoint_basicInbound
=== PAUSE TestAccAwsRoute53ResolverEndpoint_basicInbound
=== RUN   TestAccAwsRoute53ResolverEndpoint_updateOutbound
=== PAUSE TestAccAwsRoute53ResolverEndpoint_updateOutbound
=== CONT  TestAccAwsRoute53ResolverEndpoint_basicInbound
=== CONT  TestAccAwsRoute53ResolverEndpoint_updateOutbound
--- PASS: TestAccAwsRoute53ResolverEndpoint_basicInbound (101.54s)
--- PASS: TestAccAwsRoute53ResolverEndpoint_updateOutbound (551.20s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	551.218s
```

**Comment by StyleT** [comment]
We also need this feature :) Thank you guys for your hard work.

**Comment by johnjelinek** [comment]
LGTM

**Comment by kmcquade** [comment]
@paultyng - have a chance to review the recent updates?

**Comment by kmcquade** [comment]
@paultyng - Just FYI, I tested this out in our environment and it works. I didn't run into any issues.

**Comment by ewbankkit** [comment]
Rebased and re-ran acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsRoute53ResolverEndpoint_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsRoute53ResolverEndpoint_ -timeout 120m
=== RUN   TestAccAwsRoute53ResolverEndpoint_basicInbound
=== PAUSE TestAccAwsRoute53ResolverEndpoint_basicInbound
=== RUN   TestAccAwsRoute53ResolverEndpoint_updateOutbound
=== PAUSE TestAccAwsRoute53ResolverEndpoint_updateOutbound
=== CONT  TestAccAwsRoute53ResolverEndpoint_basicInbound
=== CONT  TestAccAwsRoute53ResolverEndpoint_updateOutbound
--- PASS: TestAccAwsRoute53ResolverEndpoint_basicInbound (106.17s)
--- PASS: TestAccAwsRoute53ResolverEndpoint_updateOutbound (517.93s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	517.978s
```

**Comment by nywilken** [comment]
Hi @ewbankkit thanks for rebasing, and sorry for the long lead time between my comments. I wanted to step through the implementation to help give a more informed review. 

I have a couple of small questions around the schema and I’m looking into the IP Address set/list TODOs. In quickly speaking to other maintainers the preferred is a TypeList over a TypeSet since there are computed values but I understand the issue around the non-deterministic order in the API response. I’ll follow up on this inline with your other comments. 

FYI there has been a major release since the last rebase so we will need to resolve a few things before merging. But before going down that route I would like to help come to a conclusion on the IP Address TODO.

File: aws/resource_aws_route53_resolver_endpoint.go [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+				ValidateFunc: validation.StringInSlice([]string{
+					route53resolver.ResolverEndpointDirectionInbound,
+					route53resolver.ResolverEndpointDirectionOutbound,
+				}, false),
+			},
+
+			"ip_address": {
+				Type:     schema.TypeSet,
+				Required: true,
+				MinItems: 2,
```
[the remark this answers is not in the document]
Comment (ewbankkit) [reply aws/resource_aws_route53_resolver_endpoint.go]: I'm pretty sure that in my testing I got errors return when I had just a single IP address; I'll retry.

File: aws/resource_aws_route53_resolver_endpoint.go [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"ip": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							Computed:     true,
+							ValidateFunc: validation.SingleIP(),
+						},
+						"ip_id": {
```
[the remark this answers is not in the document]
Comment (ewbankkit) [reply aws/resource_aws_route53_resolver_endpoint.go]: yes

File: aws/resource_aws_route53_resolver_endpoint.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (nywilken) [reply aws/resource_aws_route53_resolver_endpoint.go]: Sorry, I experienced the same issue.  I was just confirming that as the reason for setting MaxItems to 2. I confirmed so no need to retry on your end.

[the remark this answers is not in the document]
Comment (ewbankkit) [reply aws/resource_aws_route53_resolver_endpoint.go]: Yes, if I change `MinItems` to _1_ and attempt to create an endpoint there is an error:

```
--- FAIL: TestAccAwsRoute53ResolverEndpoint_basicInbound (26.91s)
    testing.go:538: Step 0 error: Error applying: 1 error occurred:
        	* aws_route53_resolver_endpoint.foo: 1 error occurred:
        	* aws_route53_resolver_endpoint.foo: error creating Route53 Resolver endpoint: InvalidRequestException: [RSLVR-00400] Resolver endpoint need to have at least 2 IP addresses.
        	status code: 400, request id: 41b4c038-40cf-4643-8cdf-94651a192c7d
```

[the remark this answers is not in the document]
Comment (ewbankkit) [reply aws/resource_aws_route53_resolver_endpoint.go]: The [documentation](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resolver.html#resolver-choose-vpc) says
> For each Resolver endpoint, you specify two or more IP addresses in different Availability Zones.

so

**Comment by ewbankkit** [comment]
Rebased to remove the merge conflict (and use `route53resolverconn` from the `AWSClient` structure).
Acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsRoute53ResolverEndpoint_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsRoute53ResolverEndpoint_ -timeout 120m
=== RUN   TestAccAwsRoute53ResolverEndpoint_basicInbound
=== PAUSE TestAccAwsRoute53ResolverEndpoint_basicInbound
=== RUN   TestAccAwsRoute53ResolverEndpoint_updateOutbound
=== PAUSE TestAccAwsRoute53ResolverEndpoint_updateOutbound
=== CONT  TestAccAwsRoute53ResolverEndpoint_basicInbound
=== CONT  TestAccAwsRoute53ResolverEndpoint_updateOutbound
--- PASS: TestAccAwsRoute53ResolverEndpoint_basicInbound (96.67s)
--- PASS: TestAccAwsRoute53ResolverEndpoint_updateOutbound (561.65s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	561.708s
```

**Comment by nywilken** [comment]
> Maybe we leave as I have it now - Just use `subnet_id`, which is the one `Required` field, as the hash value?
 
@ewbankkit @kmcquade do you have a sense of what the general use case would be for the subnet selection when creating resolvers? More specifically, would you say that the need is to configure resolvers with two different `subnet_ids` or is it likely that most consumers would need to define the same `subnet_id` multiple times? If it is the former that I would feel more comfortable merging this in its current state and handling the multiple subnet_id post release.

**Comment by ewbankkit** [comment]
The AWS recommendation is _two or more IP addresses in different Availability Zones_ which means distinct subnet IDs (and I think the API even enforces this, I'll check on that) and that is certainly how we will configuring our setup.

File: aws/resource_aws_route53_resolver_endpoint.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (nywilken) [reply aws/resource_aws_route53_resolver_endpoint.go]: Yeah, I agree with the documentation being incorrect at this point. Thanks for highlighting that document.

**Comment by nywilken** [comment]
> The AWS recommendation is _two or more IP addresses in different Availability Zones_ which means distinct subnet IDs (and I think the API even enforces this, I'll check on that) and that is certainly how we will configuring our setup.

I've tested locally using a TypeList and the API doesn't seem to enforce it. But I think we are in agreement on it not being a big enough use case.

```
> TF_LOG_PATH=log.txt TF_LOG=debug TF_ACC=1 go test ./aws -v -timeout 120m -parallel 10 -run=TestAccAwsRoute53ResolverEndpoint_basicInbound 
=== RUN   TestAccAwsRoute53ResolverEndpoint_basicInbound
=== PAUSE TestAccAwsRoute53ResolverEndpoint_basicInbound
=== CONT  TestAccAwsRoute53ResolverEndpoint_basicInbound
EXPAND [{
  SubnetId: "subnet-031524f87f5a1f18e"
} {
  SubnetId: "subnet-031524f87f5a1f18e"
}]
FLATTEN: [map[ip_id:rni-ea0df5ff66fa430d9 subnet_id:subnet-031524f87f5a1f18e ip:10.0.8.78] map[subnet_id:subnet-031524f87f5a1f18e ip:10.0.55.95 ip_id:rni-e638b90de437
4742a]]
FLATTEN: [map[subnet_id:subnet-031524f87f5a1f18e ip:10.0.8.78 ip_id:rni-ea0df5ff66fa430d9] map[subnet_id:subnet-031524f87f5a1f18e ip:10.0.55.95 ip_id:rni-e638b90de4374
742a]]
FLATTEN: [map[subnet_id:subnet-031524f87f5a1f18e ip:10.0.8.78 ip_id:rni-ea0df5ff66fa430d9] map[subnet_id:subnet-031524f87f5a1f18e ip:10.0.55.95 ip_id:rni-e638b90de4374
742a]]
FLATTEN: [map[ip_id:rni-ea0df5ff66fa430d9 subnet_id:subnet-031524f87f5a1f18e ip:10.0.8.78] map[subnet_id:subnet-031524f87f5a1f18e ip:10.0.55.95 ip_id:rni-e638b90de4374
742a]]
--- PASS: TestAccAwsRoute53ResolverEndpoint_basicInbound (372.50s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       372.519s

````

**Comment by ewbankkit** [comment]
Agreed, my test may have not been a real test of the API as the CLI might have collapsed the duplicated subnet IDs.

**Comment by kmcquade** [comment]
@nywilken - definitely not the same `subnet_id` multiple times... two or more different subnet IDs.

File: aws/resource_aws_route53_resolver_endpoint_test.go [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_route53_resolver_endpoint" {
+			continue
+		}
+
+		// Try to find the resource
+		_, err := conn.GetResolverEndpoint(&route53resolver.GetResolverEndpointInput{
+			ResolverEndpointId: aws.String(rs.Primary.ID),
+		})
+		if err != nil {
```
Comment (nywilken) [inline aws/resource_aws_route53_resolver_endpoint_test.go]: ditto

File: aws/resource_aws_route53_resolver_endpoint_test.go [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+		}
+
+		for _, resolverEndpoint := range page.ResolverEndpoints {
+			id := aws.StringValue(resolverEndpoint.Id)
+
+			log.Printf("[INFO] Deleting Route53 Resolver endpoint: %s", id)
+			_, err := conn.DeleteResolverEndpoint(&route53resolver.DeleteResolverEndpointInput{
+				ResolverEndpointId: aws.String(id),
+			})
+			if err != nil {
```
Comment (nywilken) [inline aws/resource_aws_route53_resolver_endpoint_test.go]: ditto

File: aws/resource_aws_route53_resolver_endpoint.go [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+
+	return nil
+}
+
+func route53ResolverEndpointRefresh(conn *route53resolver.Route53Resolver, epId string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		resp, err := conn.GetResolverEndpoint(&route53resolver.GetResolverEndpointInput{
+			ResolverEndpointId: aws.String(epId),
+		})
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (ewbankkit) [reply aws/resource_aws_route53_resolver_endpoint.go]: 👍

**Comment by ewbankkit** [comment]
Additional code review changes made.
Acceptance test:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsRoute53ResolverEndpoint_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsRoute53ResolverEndpoint_ -timeout 120m
=== RUN   TestAccAwsRoute53ResolverEndpoint_basicInbound
=== PAUSE TestAccAwsRoute53ResolverEndpoint_basicInbound
=== RUN   TestAccAwsRoute53ResolverEndpoint_updateOutbound
=== PAUSE TestAccAwsRoute53ResolverEndpoint_updateOutbound
=== CONT  TestAccAwsRoute53ResolverEndpoint_basicInbound
=== CONT  TestAccAwsRoute53ResolverEndpoint_updateOutbound
--- PASS: TestAccAwsRoute53ResolverEndpoint_basicInbound (206.00s)
--- PASS: TestAccAwsRoute53ResolverEndpoint_updateOutbound (707.33s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	707.352s
```

**Pull request merged**